src/monitoring_tests/cost_coverage_per_solver_test.py

```diff
@@ -0,0 +1,104 @@
+"""
+Computing cost coverage per solver.
+"""
+# pylint: disable=logging-fstring-interpolation
+
+from typing import Any, Dict
+from src.monitoring_tests.base_test import BaseTest
+from src.apis.web3api import Web3API
+from src.apis.orderbookapi import OrderbookAPI
+from src.constants import DAY_BLOCK_INTERVAL
+
+
+class CostCoveragePerSolverTest(BaseTest):
+    """
+    This test checks the following on a per solver basis:
+        1. fees_collected (as perceived by solver) minus actual execution cost.
+        2. total payout to solver minus actual execution cost.
+    The intent is to gain a better understanding of which solvers are more costly,
+    how much we are paying them etc.
+
+    The results are reported once a day as a log.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.web3_api = Web3API()
+        self.orderbook_api = OrderbookAPI()
+        self.cost_coverage_per_solver: Dict[str, float] = {}
+        self.total_coverage_per_solver: Dict[str, float] = {}
+        self.original_block = self.web3_api.get_current_block_number()
+
+    def cost_coverage(self, competition_data: dict[str, Any], gas_cost: float) -> bool:
+        """
+        This function compares the fees, as perceived by the winning solver, and the actual
+        execution cost of the corresponding settlement. This is refered to as cost_coverage,
+        and is supposed to monitor how well the fees end up approximating the execution cost
+        of a solution.
+        We also look at the coverage, from the protocol perspective, of each settlement, i.e.,
+        we compare the payout made to the solver with the fees collected.
+        """
+
+        solution = competition_data["solutions"][-1]
+        fees = float(int(solution["objective"]["fees"]) / 10**18)
+        surplus = float(int(solution["objective"]["surplus"]) / 10**18)
+        solver = solution["solver"]
+        ref_score = 0
+        if len(competition_data["solutions"]) > 1:
+            second_best_sol = competition_data["solutions"][-1]
```

Comment:
the second largest score should come from this solution:
```suggestion
            second_best_sol = competition_data["solutions"][-2]
```
(might not exist though)

Reply:
Oh good catch, this was a typo. Since we check the number of solutions, then we are sure it exists at this point of the code.

Reply:
Ok, the change looks good then.